src/EditorFeatures/Core/PdbSourceDocument/SourceLinkService.cs

```diff
@@ -0,0 +1,97 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.IO;
+using System.Reflection.PortableExecutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PdbSourceDocument;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.Debugger.Contracts.SourceLink;
+using Microsoft.VisualStudio.Debugger.Contracts.SymbolLocator;
+
+namespace Microsoft.CodeAnalysis.Editor.PdbSourceDocument
+{
+    [Export(typeof(ISourceLinkService)), Shared]
+    internal class SourceLinkService : ISourceLinkService
+    {
+        private readonly IDebuggerSymbolLocatorService _debuggerSymbolLocatorService;
+        private readonly IDebuggerSourceLinkService _debuggerSourceLinkService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SourceLinkService(IDebuggerSymbolLocatorService debuggerSymbolLocatorService, IDebuggerSourceLinkService debuggerSourceLinkService)
+        {
+            _debuggerSymbolLocatorService = debuggerSymbolLocatorService;
+            _debuggerSourceLinkService = debuggerSourceLinkService;
+        }
+
+        public async Task<PdbFilePathResult?> GetPdbFilePathAsync(string dllPath, PEReader peReader, IPdbSourceDocumentLogger? logger, CancellationToken cancellationToken)
+        {
+            var hasCodeViewEntry = false;
+            uint timeStamp = 0;
+            CodeViewDebugDirectoryData codeViewEntry = default;
+            using var _ = ArrayBuilder<PdbChecksum>.GetInstance(out var checksums);
+            foreach (var entry in peReader.ReadDebugDirectory())
+            {
+                if (entry.Type == DebugDirectoryEntryType.PdbChecksum)
+                {
+                    var checksum = peReader.ReadPdbChecksumDebugDirectoryData(entry);
+                    checksums.Add(new PdbChecksum(checksum.AlgorithmName, checksum.Checksum));
+                }
+                else if (entry.Type == DebugDirectoryEntryType.CodeView && entry.IsPortableCodeView)
+                {
+                    hasCodeViewEntry = true;
+                    timeStamp = entry.Stamp;
+                    codeViewEntry = peReader.ReadCodeViewDebugDirectoryData(entry);
+                }
+            }
+
+            if (!hasCodeViewEntry)
+                return null;
+
+            var pdbInfo = new SymbolLocatorPdbInfo(
+                Path.GetFileName(codeViewEntry.Path),
+                codeViewEntry.Guid,
+                (uint)codeViewEntry.Age,
+                timeStamp,
+                checksums.ToImmutable(),
+                dllPath,
+                codeViewEntry.Path);
+
+            var flags = SymbolLocatorSearchFlags.ForceMsftSymbolServer | SymbolLocatorSearchFlags.ForceNuGetSymbolServer;
```

Comment:
Why the forcing here? Does this mean even if the user turned those off it'll still use them anyways?

Reply:
Yep. By default these are both off, and so this feature is pretty useless otherwise. We talked about potentially adding a setting to control this there wasn't a big push for it.

Reply:
Going to remove this for now, until the option is in place.